### DISCUSSION TITLE
Parsing quantity strings for custom units

### DISCUSSION THREAD

**Discussion by yuri-maxiutenko** [angularsen/UnitsNet#1263] [Q&A] [answered]
Hello,

Currently I'm trying to create my custom unit to be able to parse a quantity string into an actual quantity. And looks like I couldn't find a way to do the latter.

I've added my custom IQuantity and Enum:
```
public enum ItemUnit
{
    Item,
    Dozen
}

public readonly struct Item : IQuantity
{
    public Item(double value, ItemUnit unit)
    {
        Unit = unit;
        Value = value;
    }

    //...

    Enum IQuantity.Unit => Unit;
    public ItemUnit Unit { get; }
    public QuantityValue Value { get; }

    private static readonly Item Zero = new(0, ItemUnit.Item);

    public BaseDimensions Dimensions => BaseDimensions.Dimensionless;

    public QuantityInfo QuantityInfo => new(nameof(Item), typeof(ItemUnit), new UnitInfo[]
    {
        new UnitInfo<ItemUnit>(ItemUnit.Item, "Pieces", BaseUnits.Undefined),
        new UnitInfo<ItemUnit>(ItemUnit.Dozen, "Dozens", BaseUnits.Undefined)
    }, ItemUnit.Item, Zero, BaseDimensions.Dimensionless);

    public double As(Enum unit)
    {
        return Convert.ToDouble(unit);
    }

    //...

    public IQuantity ToUnit(Enum unit)
    {
        if (unit is ItemUnit pieceUnit)
        {
            return new Item(As(unit), pieceUnit);
        }

        throw new ArgumentException("Must be of type ItemUnit.", nameof(unit));
    }

    //...

    public override string ToString()
    {
        return $"{Value} {Unit}";
    }

    public string ToString(string format, IFormatProvider formatProvider)
    {
        return $"Piece ({format}, {formatProvider})";
    }

    public string ToString(IFormatProvider provider)
    {
        return $"Piece ({provider})";
    }
}
```

Also, I've added some abbreviations:
```
UnitAbbreviationsCache.Default.MapUnitToAbbreviation(ItemUnit.Item, "item", "items", "ea", "unit", "piece", "type");
UnitAbbreviationsCache.Default.MapUnitToAbbreviation(ItemUnit.Dozen, "doz");
```

And now I want to parse a string like "6 items" to an Item object.
I've read the documentation and searched through the code, but was unable to find how to do this exactly.

The documentation clearly states, _"Reuse QuantityParser and UnitParser to parse quantity strings like "5 cm" and "cm" for your own quantities and units"_, so this should definitely be possible. But there's no code sample for this exact case there.

The usual way of using UnitsNet.Quantity.TryParse() doesn't work, because inside the TryParse method there's an explicit check if the passed type belongs to one of the types known to UnitsNet, so the custom one gets false:
```
return quantityType switch
{
       Type _ when quantityType == typeof(AbsorbedDoseOfIonizingRadiation) => /* ... */,
       Type _ when quantityType == typeof(Acceleration) =>  /* ... */,
       Type _ when quantityType == typeof(AmountOfSubstance) =>  /* ... */,
       // ...
       _ => false
}
```

I've browsed through the code and the unit tests and found exactly what I need:
```
var unitAbbreviationsCache = new UnitAbbreviationsCache();
unitAbbreviationsCache.MapUnitToAbbreviation(HowMuchUnit.Some, "foo");
var quantityParser = new QuantityParser(unitAbbreviationsCache);
HowMuch q = quantityParser.Parse<HowMuch, HowMuchUnit>("1 FOO",
    null,
    (value, unit) => new HowMuch((double) value, unit));
```

Alas, the QuantityParser class is internal, so I cannot use it directly.

Also, there's the UnitParser class which is public, but it seems to be only capable of parsing an abbreviation (e.g., "foo") to enum, which is not exactly what I need.

I have a feeling that I'm missing something simple. Could you please help me guys?

**Comment by angularsen** [comment, marked as answer]
Hi, sorry for the confusion. Custom units are quite experimental still, and has its rough edges.

`QuantityParser` was intended to be public eventually, but I guess that was forgotten.
I merged a fix now, let me know how this works out for you.
#1264

Nuget should be out shortly. 
[Release UnitsNet/5.20.0 · angularsen/UnitsNet](https://github.com/angularsen/UnitsNet/releases/tag/UnitsNet%2F5.20.0)

**Reply by yuri-maxiutenko** [reply]
I just checked it, everything works. Thanks again!

**Reply by angularsen** [reply]
I'm glad to hear it, thanks!

**Comment by angularsen** [comment]
You really made an effort to figure this out before asking too, I appreciate that 🙇

**Reply by yuri-maxiutenko** [reply]
Thanks @angularsen!
Well, your library is so good, so I had I suspicion that it's me who's doing something wrong. So I decided to investigate this issue myself before asking any dumb questions :-)
